hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/JoinRequest.java

```diff
@@ -107,6 +117,7 @@ public void readData(ObjectDataInput in) throws IOException {
 
         this.excludedMemberUuids = unmodifiableSet(excludedMemberUuids);
         this.addresses = readMap(in);
+        cpMemberUUID = UUIDSerializationUtil.readUUID(in);
```

Comment:
Would this create issue with 5.1 and 5.2 mixed clusters?

Reply:
`JoinRequest` is a bit special because at the time it is prepared and sent, cluster version is still `UNKNOWN` (cluster version is set by master in `FinalizeJoinOp`), so the regular `Versioned` serialization cannot apply to it.
We should use the `memberVersion` that is already a part of `JoinRequest` (inherited from `JoinMessage` and already read in `super.readData`) eg:

```
        // RU_COMPAT_5_1
        MemberVersion current = MemberVersion.of(5, 2, 0);
        if (MemberVersion.MAJOR_MINOR_VERSION_COMPARATOR.compare(this.memberVersion, current) >= 0) {
            // member is at least 5.2, we expect cpMemberUuid exists in the incoming packet
        }
```

On the `writeData` side we can safely write the new `cpMemberUuid` field unconditionally at the end of the message. A 5.1 reader will safely skip it since the new bytes are written at the end of the packet and `JoinRequest` is not subclassed so we don't mess with the expected position of fields further down the byte stream.